piff/knn_interp.py

```diff
@@ -0,0 +1,373 @@
+        """
+        return numpy.array([star.fit[key] for key in self.attr_target])
+
+    def initialize(self, star_list, logger=None):
+        """Solve for the interpolation coefficients given stars and attributes
+
+        :param star_list:   A list of Star instances to interpolate between
+        :param logger:      A logger object for logging debug info. [default: None]
+        """
+        self.solve(star_list, logger=logger)
+
+    def solve(self, star_list, logger=None):
+        """Solve for the interpolation coefficients given stars and attributes
+
+        :param star_list:   A list of Star instances to interpolate between
+        :param logger:      A logger object for logging debug info. [default: None]
+        """
+        X = numpy.array([self.getProperties(star) for star in star_list])
+        y = numpy.array([self.getFitProperties(star) for star in star_list])
+        self._fit(X, y)
+
+    def interpolate(self, star, logger=None):
+        """Perform the interpolation to find the interpolated parameter vector at some position. Calls interpolateList because sklearn prefers list input anyways
+
+        :param star:        A Star instance to which one wants to interpolate
+        :param logger:      A logger object for logging debug info. [default: None]
+
+        :returns: a new Star instance with its StarFit member holding the interpolated parameters
+        """
+        # because of sklearn formatting, call interpolateList and take 0th entry
+        return self.interpolateList([star], logger=logger)[0]
+
+    def interpolateList(self, star_list, logger=None):
+        """Perform the interpolation for a list of stars.
+
+        :param star_list:   A list of Star instances to which to interpolate.
+        :param logger:      A logger object for logging debug info. [default: None]
+
+        :returns: a list of new Star instances with interpolated parameters
+        """
+
+        X = numpy.array([self.getProperties(star) for star in star_list])
+        # y = numpy.array([self.getFitProperties(star) for star in star_list])
+        y = self._predict(X)
+        star_list_fitted = []
+        for yi, star in zip(y, star_list):
+            if star.fit is None:
+                fit = StarFit(yi)
+            else:
+                try:
+                    fit = star.fit.newParams(yi)
+                except TypeError:
+                    if logger:
+                        logger.info('Warning, stars interpolated to fewer params! %s', len(fit))
+                    fit = StarFit(yi)
+            star_list_fitted.append(Star(star.data, fit))
+        return star_list_fitted
+
+    def writeSolution(self, fits, extname):
+        """Write the solution to a FITS binary table.
+
+        Save the knn params and the X and y arrays
+
+        :param fits:        An open fitsio.FITS object.
+        :param extname:     The name of the extension with the interp information.
+        """
```

Comment:
I'm following here the example of polynomial_interp.py, or at least I think I am. I can understand why attr_interp and attr_target could be considered kwargs, but I think in the not-#20 branch, kwargs couldn't handle writing out lists or arrays, which is why they got lumped this way. If that has changed, I'll gladly put them into the kwargs dictionary.

If the X and y (now locations and targets) need to be a part of the kwargs, then I am confused by what the kwargs are supposed to be. These are the solution to the interpolator, just like the coefficients in the polynomial interp and the mean value in mean interp.